### PR TITLE
fix(core): interval resolution should upcast to smallest unit

### LINKED ISSUE #6139: bug: interval addition/subtraction should upcast to smallest unit
### What happened?

Currently adding/subtracting intervals results in an interval with a unit of the `left` operand. This can result in accidental information loss if the smaller unit in the operation is the right operand:

```python
In [1]: import ibis

In [2]: one_second = ibis.interval(seconds=1)

In [3]: two_minutes = ibis.interval(minutes=2)

In [4]: (one_second + two_minutes).type().unit
Out[4]: <IntervalUnit.SECOND: 's'>

In [5]: (two_minutes + one_second).type().unit  # I'd expect seconds here too
Out[5]: <IntervalUnit.MINUTE: 'm'>

In [6]: (one_second + two_minutes).execute()
Out[6]: Timedelta('0 days 00:02:01')

In [7]: (two_minutes + one_second).execute()  # the second is dropped on rounding to minutes
Out[7]: Timedelta('0 days 00:02:00')
```

The offending code is [here](https://github.com/ibis-project/ibis/blob/5a013ffd3032e596a16a6a71476c2754f9bc9c8b/ibis/expr/operations/temporal.py#L297). With some pointers I can probably fix this. I know we have casting utilities for handling integer upcasting, but I'm not sure if the same exists for intervals.

### What version of ibis are you using?

dev

### What backend(s) are you using, if any?

_No response_

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by mesejo**
Inspired by pandas [resolution_string](https://github.com/pandas-dev/pandas/blob/e19ceb76a4c5903469fea93040c174633c84b9b9/pandas/_libs/tslibs/timedeltas.pyx#L1383)

fixes #6139

### CODE REVIEW

File: ibis/expr/rules.py [hunk 1]
```diff
@@ -432,6 +433,31 @@ def output_dtype(self):
...
+    elif IntervalUnit.DAY in units:
+        return IntervalUnit.DAY
+    elif IntervalUnit.WEEK in units:
+        return IntervalUnit.WEEK
+    elif IntervalUnit.MONTH in units:
+        return IntervalUnit.MONTH
+    elif IntervalUnit.QUARTER in units:
+        return IntervalUnit.QUARTER
+    else:
+        return IntervalUnit.YEAR
```
Comment (jcrist) [inline ibis/expr/rules.py:458]:
```suggestion
    # Find the smallest unit present in units
    for unit in reversed(IntervalUnit):
        if unit in units:
            return unit
    assert False, "unreachable"
```

Comment (jcrist) [inline ibis/expr/rules.py:458]: The following might be a bit cleaner:

```suggestion
    # Find the smallest unit present in units
    for unit in reversed(IntervalUnit):
        if unit in units:
            return unit
    assert False, "unreachable"
```

**Review by jcrist — COMMENTED** [review]
Overall this looks good to me! Just one small fixup.

**Comment by kszucs** [comment]
Could you please add a `backends` test case to `ibis/backends/tests/test_temporal.py`, since some of the backends may not support operations with operands using different units?

File: ibis/expr/rules.py (hunk 1 above)
Comment (mesejo) [reply ibis/expr/rules.py]: Thanks for the suggestion!

**Comment by mesejo** [comment]
> Could you please add a backends test case to ibis/backends/tests/test_temporal.py, since some of the backends may not support operations with operands using different units?

@kszucs Thanks for the feedback. I added a test case [here](https://github.com/ibis-project/ibis/pull/6203/files#diff-561fbc13e95852435279b84b20e04d88a399427943267a0cd50d59d00e839636R833) Do you think it is enough?

**Comment by jcrist** [comment]
Yep, looks good to me! Can you squash your commits together, and ensure the commit message follows our [commit message requirements](https://ibis-project.org/community/contribute/02_workflow/#submit-a-pr) (this is what caused the failing test)? After that this should be good to merge.

**Review by kszucs — APPROVED** [review]
LGTM, thanks @mesejo!

**Pull request merged**